source/client_channel_handler.c

```diff
@@ -626,6 +626,7 @@ void aws_mqtt_internal_complete_request(struct aws_mqtt_request *request, int er
     /* Fire the callback and clean up the memory, as the connection get destoried. */
     if (!request->completed) {
         if (request->on_complete) {
+            /* TODO: I am invoking the callback with lock hold XXXXXX */
             request->on_complete(connection, request->packet_id, error_code, request->on_complete_ud);
```

Comment:
Is it safe to do the following?
```
mqtt_connection_unlock_synced_data(connection);
request->on_complete(connection, request->packet_id, error_code, request->on_complete_ud);
mqtt_connection_lock_synced_data(connection);
```

Reply:
NOPE, CANNOT DO THIS! [It](https://github.com/awslabs/aws-c-mqtt/pull/157/checks?check_run_id=1233344391) almost killed me!!!